ext/cl_khr_integer_dot_product.asciidoc

```diff
@@ -0,0 +1,212 @@
+Version 3.0.6.
+
+This extension requires OpenCL 1.0.
+
+==== Contributors
+
+Kévin Petit, Arm Ltd. +
+Jeremy Kemp, Imagination Technologies +
+Ben Ashbaugh, Intel +
+Ruihao Zhang, Qualcomm +
+Stuart Brady, Arm Ltd +
+Balaji Calidas, Qualcomm +
+Ayal Zaks, Intel +
+
+=== New API Enums
+
+Accepted value for the _param_name_ parameter to *clGetDeviceInfo*:
+
+[source,c]
+----
+
+CL_DEVICE_INTEGER_DOT_PRODUCT_INPUT_4x8BIT_PACKED_KHR      (1 << 0)
+CL_DEVICE_INTEGER_DOT_PRODUCT_INPUT_4x8BIT_KHR             (1 << 1)
+
+CL_DEVICE_INTEGER_DOT_PRODUCT_CAPABILITIES_KHR              0x1073
+----
+
+=== New OpenCL C Functions
+
+This extension defines a number of new functions that operate on vectors
+of integers. The exact function overloads available depend on the features
+supported by the target device.
+
+[source,c]
+----
+uint dot(uchar4 a, uchar4 b);
+int dot(char4 a, char4 b);
+int dot(uchar4 a, char4 b);
+int dot(char4 a, uchar4 b);
+
+uint dot_acc_sat(uchar4 a, uchar4 b, uint acc);
+int dot_acc_sat(char4 a, char4 b, int acc);
+int dot_acc_sat(uchar4 a, char4 b, int acc);
+int dot_acc_sat(char4 a, uchar4 b, int acc);
+
+uint dot_4x8packed_uu_uint(uint a, uint b);
+int dot_4x8packed_ss_int(uint a, uint b);
+int dot_4x8packed_us_int(uint a, uint b);
+int dot_4x8packed_su_int(uint a, uint b);
+
+uint dot_acc_sat_4x8packed_uu_uint(uint a, uint b, uint acc);
+int dot_acc_sat_4x8packed_ss_int(uint a, uint b, int acc);
+int dot_acc_sat_4x8packed_us_int(uint a, uint b, int acc);
+int dot_acc_sat_4x8packed_su_int(uint a, uint b, int acc);
+----
+
+=== Modifications to the OpenCL API Specification
+
+(Modify Section 4.2, *Querying Devices*) ::
++
+--
+
+(Add the following to Table 4.3, _Device Queries_) ::
++
+--
+
+[cols="2,2,4",options="header"]
+|====
+| cl_device_info
+| Return Type
+| Description
+
+| `CL_DEVICE_INTEGER_DOT_PRODUCT_CAPABILITIES_KHR`
+| `cl_device_integer_dot_product_capabilities_khr`
+| Returns the integer dot product capabilities supported by the device. +
+ +
+`CL_DEVICE_INTEGER_DOT_PRODUCT_INPUT_4x8BIT_PACKED_KHR` is always set
+  indicating that all implementations that support `cl_khr_integer_dot_product`
+  must support dot product built-in functions and, when SPIR-V is supported,
+  SPIR-V instructions that take four-component vectors of 8-bit integers packed
+  into 32-bit integers as input. +
+`CL_DEVICE_INTEGER_DOT_PRODUCT_INPUT_4x8BIT_KHR` is set when dot product
+  built-in functions and, when SPIR-V is supported, SPIR-V instructions that
+  take four-component of 8-bit elements as input are supported. +
+
+|====
+
+--
+OpenCL 3 devices must report the following feature macros via
+`CL_DEVICE_OPENCL_C_FEATURES` when the corresponding bit is set in the bitfield
+returned for `CL_DEVICE_INTEGER_DOT_PRODUCT_CAPABILITIES_KHR`:
+
+[cols="1,1",options="header"]
+|====
+| Feature bit
+| Feature macro
+
+| `CL_DEVICE_INTEGER_DOT_PRODUCT_INPUT_4x8BIT_KHR`
+| `__opencl_c_integer_dot_product_input_4x8bit`
```

Comment:
I think this is fine, but just to confirm: we don't want a feature test macro for the packed variants because they are required and the test for the extension is sufficient?

Would we want to add a feature test macro if this extension ever became a core feature?

We don't want a `_khr` suffix on this feature test macro?

Reply:
>we don't want a feature test macro for the packed variants because they are required and the test for the extension is sufficient?

That's been the rationale.
>Would we want to add a feature test macro if this extension ever became a core feature?

Probably, if we make it an optional core feature. There could be an argument for adding it now to help with code portability if we make this a core feature in the future.

>We don't want a _khr suffix on this feature test macro?

I'd say no for the same reason we don't add it to khr built-in function names. We want code written for the extension to "just work" if it becomes a core feature.

Reply:
As discussed in the 2021/06/22 teleconference, I've added a feature macro for `CL_DEVICE_INTEGER_DOT_PRODUCT_INPUT_4x8BIT_PACKED_KHR`